### PR TITLE
Add the option for inactive top cells

### PR DESCRIPTION
**Pull request by cbegeman**
Inactive top cells are introduced to add flexibility to the vertical coordinate below ice shelves. 

This PR introduces phase 1 in the development documented in https://github.com/MPAS-Dev/MPAS-Model/pull/811, which touches only the code needed to run a basic ice shelf test case.

The variable `minLevelCell` and related `minLevelEdge*` and `minLevelVertex*` variables are introduced by analogy with `maxLevel*` variables. By default, there are no inactive top cells and `minLevel*` = 1. The main code changes are to k-loop limits and designating surface quantities as those at `minLevel*`.

### CODE REVIEW

**Comment by xylar** [comment]
## Testing

(Edit this comment to add testing as it is performed).

### COMPASS

As a baseline in COMPASS, I'm using the `legacy` branch + https://github.com/xylar/compass/commit/ca592c778ea4144efe4db02c025c8e5ae0be8ca9, which adds a test suite (`z_level_dev`) that includes both all tests in the `nightly` suite and several tests from `land_ice_fluxes`.

The COMPASS branch used for testing is https://github.com/xylar/compass/tree/add_z_level_sub_ice_shelf_2D, but I no COMPASS changes should be required so testing with `legacy` should also work. (I will verify this.)

The results from the `z_level_dev` test suite are bit-for-bit identical to the baseline for:
* Ubuntu 20.04 and Gnu 9.3.0
* Anvil and Intel
* ...

### E3SM

...

**Comment by cbegeman** [comment]
> The results from the z_level_dev test suite are bit-for-bit identical to the baseline for:
* Grizzly and Intel

**Comment by xylar** [comment]
@cbegeman, as I noted above, a corresponding PR into `legacy` in the compass repo is apparently going to be required.  I thought `legacy` might just work but apparently some of my changes are needed.

**Comment by cbegeman** [comment]
> @cbegeman, could you re-test on Grizzly with the new COMPASS branch I created? I'll re-test on Anvil.

Yes, we get bfb with the new COMPASS branch on Grizzly.

**Comment by xylar** [comment]
@cbegeman, E3SM tests vs. baseline are failing.  Here are some details.

When I do a test merge at the following commits, I am indicating which pass and which fail `SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel`

Commits failing:
* f8819694 Set vertical velocity top to 0 from k=1:minLevelCell-1

Commits passing:
* 73826a31 add minLevel* to ocn_vel_pressure_grad_tend

I will continue updating as new test results come in.

**Comment by cbegeman** [comment]
@xylar Thanks for tracking this down. c85f36c was the commit I introduced so that QU240-Analysis was bfb on Grizzly. I'll remove it and test again there.

**Comment by xylar** [comment]
If the QU240 analysis is failing without this, I wonder if it's also failing with ocean/develop. It's hard to see how this could be caused by our changes.

**Comment by cbegeman** [comment]
So QU240-Analysis is now passing with MPAS-Dev/compass#55, which suggests that the difference is due to the xylar/compass/add_z_level_sub_ice_shelf_2D commits. I'll retest this branch without commit c85f36c with xylar/compass/add_z_level_sub_ice_shelf_2D.

**Comment by xylar** [comment]
If it's still failing, let me know. Something to fix for phase 2.

**Comment by xylar** [comment]
@cbegeman, I haven't had a chance yet but I'll retest everything on Anvil soon, then try again on Compy.

File: src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F
Comment (mattdturner) [inline src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:1750]: For this loop, if `minLevelCell(iCell)` is set to 1, then you are looping from `1` to `1+config_reset_debugTracers_top_nLayers`.  Previously, the loop was from `1` to `config_reset_debugTracers_top_nLayers`.  Is the desired behavior to change the number of iterations of this loop for the case when `minLevelCell(iCell) == 1`?

Comment (cbegeman) [reply src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:1750]: Good catch. No, we do want the same number of iterations:
```suggestion
                           do k = minLevelCell(iCell), minLevelCell(iCell)+config_reset_debugTracers_top_nLayers-1
```

File: src/core_ocean/shared/mpas_ocn_diagnostics.F
Comment (mattdturner) [inline src/core_ocean/shared/mpas_ocn_diagnostics.F:2152]: Should this line be changed to:
```
velocityMagnitude = sqrt(kineticEnergyCell(minLevelCell(cell1),cell1) + kineticEnergyCell(minLevelCell(cell2),cell2))
```

In other words, should the 2nd `minLevelCell(cell1)` be changed to `minLevelCell(cell2)` to match the `cell2` in the 2nd index?

Comment (cbegeman) [reply src/core_ocean/shared/mpas_ocn_diagnostics.F:2152]: Totally. Thanks!
```suggestion
         velocityMagnitude = sqrt(kineticEnergyCell(minLevelCell(cell1),cell1) + kineticEnergyCell(minLevelCell(cell2),cell2))
```

File: src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F
Comment (mattdturner) [inline src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:1762]: The upper limit of this loop should also have the `-1` appended (like line 1750)

File: src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F
Comment (mattdturner) [inline src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:1766]: The upper limit of this loop should also have the `-1` appended (like line 1750)

File: src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F
Comment (mattdturner) [inline src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:1780]: The upper limit of this loop should also have the `-1` appended (like line 1750)

File: src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F
Comment (mattdturner) [inline src/core_ocean/mode_forward/mpas_ocn_time_integration_split.F:1784]: The upper limit of this loop should also have the `-1` appended (like line 1750)

**Comment by mattdturner** [comment]
@cbegeman Thanks for the work in this PR.  I added a few additional comments to the split explicit driver regarding loop limits.  

Aside from those comments, I think this looks good by visual inspection.  Once those loop limits are updated, I will approve this PR and rely on testing by you and @xylar

**Comment by cbegeman** [comment]
@mattdturner Thanks for your thorough review! I got those final loop limit changes to split-explicit in.

@xylar and I are working on E3SM tests on Cori, Anvil and Compy so we'll keep you posted on that.

**Comment by xylar** [comment]
@cbegeman, I'm seeing DIFFs in `SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi`

I'll try rolling back commits to figure out where it's happening.

**Comment by xylar** [comment]
@cbegeman, I haven't seen this issue in compass testing, but I've never run compass with PGI.  I'm pretty close to having the first commit that causes diffs identified.  I'll let you know.

**Comment by cbegeman** [comment]
@xylar Thanks for looking into it. My standalone z-level-dev tests pass (Grizzly, intel).

**Comment by xylar** [comment]
The (first) problem commit is:
```
73826a31 add minLevel* to ocn_vel_pressure_grad_tend
```

**Comment by xylar** [comment]
@cbegeman, some very good news.  I rebased all commits except 73826a3 onto 95daa3f (which was the submodule pointed to by components/mpas-source when I started testing today).  That passed the `SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi` test without any diffs.  So it's just that one commit that I have to debug.

**Comment by mark-petersen** [comment]
@cbegeman should we give this a final review and shoot for a Monday merge? Are there any unresolved issues?  We just merged https://github.com/MPAS-Dev/MPAS-Model/pull/826 yesterday - there will be a few lines in conflict with this PR. Please rebase on `ocean/develop`.

**Comment by xylar** [comment]
@mark-petersen, the diffs I'm seeing with PGI on Compy are still unresolved.  The ball is in my court and I haven't made any headway on this today.  Hopefully tomorrow.

**Comment by xylar** [comment]
> We just merged #826 yesterday - there will be a few lines in conflict with this PR. Please rebase on ocean/develop.

We already did that yesterday.

**Comment by xylar** [comment]
@cbegeman, as I noted above, I'm getting bit-for-bit results with:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
```
(that is, with PGI in debug mode) but not when it's optimized.  I think we have to live with the non-bit-for-bit optimized build on Compy because I don't see any way to trick the compiler.  Presumably, it's optimizing something with the old code:
```
do k=2,kMax
```
that can't be optimized with the new code:
```
do k=kMin+1,kMax
```

**Review by xylar — APPROVED** [review]
I think this is good to go.  I believe @cbegeman mentioned one last rebase she wants to do, and I'll rerun my tests after that, but the only diffs I'm seeing are those noted for compy with optimized (but not debug) PGI.

**Comment by cbegeman** [comment]
@xylar Thanks for doing that testing. It seems reasonable to me to proceed. 

It looks like commit 8329b9c is the last commit needed after rebasing onto `ocean/develop` after @mattdturner's changes to split-explicit. Did you test with this commit?

**Comment by xylar** [comment]
> It looks like commit 8329b9c is the last commit needed after rebasing onto ocean/develop after @mattdturner's changes to split-explicit. Did you test with this commit?

I think so but I'm not 100% sure.  Best to just re-test tomorrow.

**Comment by cbegeman** [comment]
MPAS standalone tests pass (z-level-dev suite, which includes nightly suite).

**The following cases pass on Cori:**
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
[To be continued]

**Comment by mark-petersen** [comment]
@cbegeman and @xylar If those test go well, are you comfortable proceeding? I can make the E3SM PR, and I leave this MPAS PR open for the day.  If everything goes well, we would merge the E3SM PR into `next` today, and make the final merge for both MPAS and E3SM in tomorrow.

**Comment by cbegeman** [comment]
@mark-petersen That sounds great to me. Thanks!

**Comment by xylar** [comment]
@mark-petersen, yes, please go ahead with the E3SM PR.

My tests on Anvil are done (all pass).  My testing on Compy is taking quite a long time.  I will keep you all posted.

**Comment by cbegeman** [comment]
@mark-petersen I just want to remind you of this PR https://github.com/MPAS-Dev/compass/pull/55 to accompany this one.

**Review by mark-petersen — APPROVED** [review]
Tested with gnu 5.3 and intel 19 on grizzly, optimized and debug, on nightly regression suite. It is BFB with current `ocean/develop` branch in optimized for both. Tested with legacy compass and compass 1.0. Thanks!

This is too much code to review visually, but I was involved with the design document discussions, so will rely on that here.

**Review by mattdturner — APPROVED** [review]
Approved via visual inspection and testing by @cbegeman and @xylar

**Comment by xylar** [comment]
I am still working on 3 (slow) tests on Compy.  I will update my "Testing" comment and post the list of completed tests on https://github.com/E3SM-Project/E3SM/pull/4171 once they are done.

**Comment by mark-petersen** [comment]
@dengwirda and @philipwjones please approve. We hope to merge today or tomorrow.

**Comment by philipwjones** [comment]
@mattdturner @cbegeman @xylar Has anyone looked at the performance/timing diffs for these changes in any of the tests with optimization on?

**Comment by xylar** [comment]
> @mattdturner @cbegeman @xylar Has anyone looked at the performance/timing diffs for these changes in any of the tests with optimization on?

@philipwjones, I don't believe so.  We were not given a clear idea of what we needed to be looking at in which tests.  This is a bit part of what we were requesting your help with both in your review here and in the design doc.

**Comment by mattdturner** [comment]
@philipwjones I just ran a quick test on Chrysalis.

Test case: `SMS_P2560x1.T62_oRRS18to6v3.GMPAS-IAF.chrysalis_intel`

Before these changes:
```
    OCN Run Time:     698.355 seconds      139.671 seconds/mday         1.69 myears/wday
```

After these changes are applied:
```
    OCN Run Time:     702.525 seconds      140.505 seconds/mday         1.68 myears/wday                    
```

**Comment by cbegeman** [comment]
@philipwjones Timing differences between the runs appear to be in the noise:

Test case: `SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel`, 1024 proc.

Before these changes:
`OCN Run Time:    3388.555 seconds      677.711 seconds/mday         0.35 myears/wday`
After these changes:
`OCN Run Time:    3378.931 seconds      675.786 seconds/mday         0.35 myears/wday`

Test case: `PEM_Ln9_P1024.ne30_oECv3.A_WCYCL1850S.cori-knl_intel`

Before these changes:
`OCN Run Time:       5.651 seconds       30.138 seconds/mday         7.85 myears/wday`
After these changes:
`OCN Run Time:       5.618 seconds       29.964 seconds/mday         7.90 myears/wday`

**Comment by philipwjones** [comment]
Thanks @cbegeman and @mattdturner - I also ended up doing a couple of quick tests. PGI on Summit may be slightly faster but in the noise. Intel/grizzly slightly slower at 1-2% level but Intel tends to do a better job of vectorizing so probably more impacted by this change and not surprising.

**Review by philipwjones — APPROVED** [review]
Approve based on testing by myself and others, prior discussions on design, and visual inspection of a few routines.

**Comment by xylar** [comment]
Thanks for the testing, @philipwjones!  That's helpful.  I'm also not surprised to see a little slowdown.  I was guessing that PGI might show some (especially because it's the only compiler where we're seeing non-bit-for-bit results for the optimized code).  I'll see what timings look like when my last run on Compy finishes.

File: src/core_ocean/shared/mpas_ocn_vmix.F
Comment (dengwirda) [inline src/core_ocean/shared/mpas_ocn_vmix.F:959]: Is there a slightly different treatment of zeroing-out unneeded entries in the tridiagonals between `tracer_vmix_tend_implicit` and `vel_vmix_tend_implicit`?

It looks like `A(1:Nsurf) = 0.0`, etc is done in `vel_vmix`, but only `A(Nsurf) = 0.0` in `tracer_vmix`. I think  there may only be a difference in unused values that are out of range, so I don't think this will impact results in any way.

**Review by dengwirda — APPROVED** [review]
Approved based on visual inspection + compiling/running locally.

File: src/core_ocean/shared/mpas_ocn_vmix.F
Comment (cbegeman) [reply src/core_ocean/shared/mpas_ocn_vmix.F:959]: You're right. I didn't catch that because it is unused but we should be consistent. I'm happy to commit this change, but I might include it in phase 2 if including it here would hold up the merge since our tests are pretty much complete. @xylar or @mark-petersen do you have opinions about whether to commit this now or later?
```suggestion
         A(1:Nsurf)=0.0_RKIND
```

**Comment by cbegeman** [comment]
@dengwirda Thanks for your review!

File: src/core_ocean/shared/mpas_ocn_vmix.F
Comment (dengwirda) [reply src/core_ocean/shared/mpas_ocn_vmix.F:959]: I'm personally not worried about this, and think it can go into phase II --- I think the only difference would be that some array entries (that are never used) could be allocated with undefined values rather than being zeroed out.
 
This shouldn't happen in the `minLevelCell=1` case anyway though, as `A(1:Nsurf)` and `A(Nsurf)`, etc should do the same zeroing.

**Comment by xylar** [comment]
## Timing on Compy

### `PET_Ln9_P1024.ne30_oECv3.A_WCYCL1850S.compy_intel`

`E3SM/master`:
```
OCN Run Time:       1.643 seconds        8.760 seconds/mday        27.02 myears/wday
```

with this branch:
```
OCN Run Time:       1.149 seconds        6.126 seconds/mday        38.64 myears/wday
```

### `SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel`

`E3SM/master`:
```
OCN Run Time:      40.649 seconds        8.130 seconds/mday        29.12 myears/wday
```

with this branch:
```
OCN Run Time:      41.177 seconds        8.235 seconds/mday        28.74 myears/wday
```

### `SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi`

`E3SM/master`:
```
OCN Run Time:      42.600 seconds        8.520 seconds/mday        27.78 myears/wday 
```

with this branch:
```
OCN Run Time:      43.195 seconds        8.639 seconds/mday        27.40 myears/wday
```

**Comment by xylar** [comment]
Thanks so much for the hard work on this, @cbegeman!  Feel free to delete you branch and any others lying around from your testing.

**Pull request merged**
